File: workout.go
// workout.go - Дневник тренировок на Go (CLI)
package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"
)

type Workout struct {
	ID       int     `json:"id"`
	Exercise string  `json:"exercise"`
	Sets     int     `json:"sets"`
	Reps     int     `json:"reps"`
	Weight   float64 `json:"weight"`
	Date     string  `json:"date"`
	Notes    string  `json:"notes"`
}

type Tracker struct {
	Workouts []Workout `json:"workouts"`
	NextID   int       `json:"next_id"`
}

const dataFile = "workouts.json"

func loadTracker() *Tracker {
	var t Tracker
	file, err := os.ReadFile(dataFile)
	if err != nil {
		t.Workouts = []Workout{}
		t.NextID = 1
		return &t
	}
	err = json.Unmarshal(file, &t)
	if err != nil {
		t.Workouts = []Workout{}
		t.NextID = 1
	}
	return &t
}

func saveTracker(t *Tracker) {
	data, _ := json.MarshalIndent(t, "", "  ")
	os.WriteFile(dataFile, data, 0644)
}

func addWorkout(t *Tracker, exercise string, sets int, reps int, weight float64, dateStr string, notes string) Workout {
	if dateStr == "" {
		dateStr = time.Now().Format("2006-01-02")
	}
	w := Workout{
		ID:       t.NextID,
		Exercise: exercise,
		Sets:     sets,
		Reps:     reps,
		Weight:   weight,
		Date:     dateStr,
		Notes:    notes,
	}
	t.Workouts = append(t.Workouts, w)
	t.NextID++
	saveTracker(t)
	return w
}

func deleteWorkout(t *Tracker, id int) bool {
	for i, w := range t.Workouts {
		if w.ID == id {
			t.Workouts = append(t.Workouts[:i], t.Workouts[i+1:]...)
			saveTracker(t)
			return true
		}
	}
	return false
}

func getWorkouts(t *Tracker, exercise string, dateFrom, dateTo string, minWeight, maxWeight float64) []Workout {
	result := t.Workouts
	if exercise != "" {
		result = filterByExercise(result, exercise)
	}
	if dateFrom != "" {
		result = filterByDateFrom(result, dateFrom)
	}
	if dateTo != "" {
		result = filterByDateTo(result, dateTo)
	}
	if minWeight > 0 {
		result = filterByMinWeight(result, minWeight)
	}
	if maxWeight > 0 {
		result = filterByMaxWeight(result, maxWeight)
	}
	sortByDate(result)
	return result
}

func filterByExercise(ws []Workout, ex string) []Workout {
	var res []Workout
	for _, w := range ws {
		if strings.EqualFold(w.Exercise, ex) {
			res = append(res, w)
		}
	}
	return res
}
func filterByDateFrom(ws []Workout, from string) []Workout {
	var res []Workout
	for _, w := range ws {
		if w.Date >= from {
			res = append(res, w)
		}
	}
	return res
}
func filterByDateTo(ws []Workout, to string) []Workout {
	var res []Workout
	for _, w := range ws {
		if w.Date <= to {
			res = append(res, w)
		}
	}
	return res
}
func filterByMinWeight(ws []Workout, min float64) []Workout {
	var res []Workout
	for _, w := range ws {
		if w.Weight >= min {
			res = append(res, w)
		}
	}
	return res
}
func filterByMaxWeight(ws []Workout, max float64) []Workout {
	var res []Workout
	for _, w := range ws {
		if w.Weight <= max {
			res = append(res, w)
		}
	}
	return res
}
func sortByDate(ws []Workout) {
	for i := 0; i < len(ws); i++ {
		for j := i + 1; j < len(ws); j++ {
			if ws[i].Date > ws[j].Date {
				ws[i], ws[j] = ws[j], ws[i]
			}
		}
	}
}

func getStatistics(t *Tracker, exercise string) (totalWorkouts int, totalVolume float64, bestByExercise map[string]struct{ Workout Workout; Volume float64 }, progress map[string]float64) {
	workouts := getWorkouts(t, exercise, "", "", 0, 0)
	totalWorkouts = len(workouts)
	totalVolume = 0
	bestByExercise = make(map[string]struct{ Workout Workout; Volume float64 })
	for _, w := range workouts {
		vol := float64(w.Sets * w.Reps) * w.Weight
		totalVolume += vol
		if best, ok := bestByExercise[w.Exercise]; !ok || vol > best.Volume {
			bestByExercise[w.Exercise] = struct{ Workout Workout; Volume float64 }{Workout: w, Volume: vol}
		}
	}
	progress = make(map[string]float64)
	if exercise != "" {
		monthMap := make(map[string]struct{ total float64; count int })
		for _, w := range workouts {
			month := w.Date[:7]
			entry := monthMap[month]
			entry.total += w.Weight
			entry.count++
			monthMap[month] = entry
		}
		// сортируем месяцы
		var months []string
		for m := range monthMap {
			months = append(months, m)
		}
		sort.Strings(months)
		for _, m := range months {
			progress[m] = monthMap[m].total / float64(monthMap[m].count)
		}
	}
	return
}

func getPersonalRecords(t *Tracker) map[string]Workout {
	best := make(map[string]Workout)
	for _, w := range t.Workouts {
		vol := float64(w.Sets*w.Reps) * w.Weight
		if existing, ok := best[w.Exercise]; !ok || vol > float64(existing.Sets*existing.Reps)*existing.Weight {
			best[w.Exercise] = w
		}
	}
	return best
}

func exportCSV(t *Tracker, filepath string) {
	file, err := os.Create(filepath)
	if err != nil {
		fmt.Println("Ошибка создания файла:", err)
		return
	}
	defer file.Close()
	writer := csv.NewWriter(file)
	defer writer.Flush()
	writer.Write([]string{"ID", "Exercise", "Sets", "Reps", "Weight", "Date", "Notes"})
	for _, w := range t.Workouts {
		writer.Write([]string{
			strconv.Itoa(w.ID),
			w.Exercise,
			strconv.Itoa(w.Sets),
			strconv.Itoa(w.Reps),
			fmt.Sprintf("%.1f", w.Weight),
			w.Date,
			w.Notes,
		})
	}
}

func main() {
	var (
		cmd       string
		exercise  string
		sets      int
		reps      int
		weight    float64
		dateStr   string
		notes     string
		id        int
		from      string
		to        string
		minWeight float64
		maxWeight float64
		output    string
	)
	flag.StringVar(&cmd, "cmd", "", "Команда: add, list, stats, pr, delete, export")
	flag.StringVar(&exercise, "exercise", "", "Упражнение")
	flag.IntVar(&sets, "sets", 0, "Подходы")
	flag.IntVar(&reps, "reps", 0, "Повторения")
	flag.Float64Var(&weight, "weight", 0, "Вес (кг)")
	flag.StringVar(&dateStr, "date", "", "Дата")
	flag.StringVar(&notes, "notes", "", "Заметки")
	flag.IntVar(&id, "id", 0, "ID записи")
	flag.StringVar(&from, "from", "", "Дата от")
	flag.StringVar(&to, "to", "", "Дата до")
	flag.Float64Var(&minWeight, "min-weight", 0, "Мин. вес")
	flag.Float64Var(&maxWeight, "max-weight", 0, "Макс. вес")
	flag.StringVar(&output, "output", "", "Файл для экспорта CSV")
	flag.Parse()

	tracker := loadTracker()

	switch cmd {
	case "add":
		if exercise == "" || sets == 0 || reps == 0 || weight == 0 {
			fmt.Println("Укажите --exercise, --sets, --reps, --weight")
			return
		}
		w := addWorkout(tracker, exercise, sets, reps, weight, dateStr, notes)
		fmt.Printf("✅ Тренировка #%d добавлена: %s %d×%d×%.1fкг\n", w.ID, w.Exercise, w.Sets, w.Reps, w.Weight)
	case "list":
		workouts := getWorkouts(tracker, exercise, from, to, minWeight, maxWeight)
		if len(workouts) == 0 {
			fmt.Println("Нет записей.")
		} else {
			fmt.Printf("%-4s %-15s %-8s %-6s %-6s %-12s %s\n", "ID", "Упражнение", "Подходы", "Повт.", "Вес", "Дата", "Заметки")
			for _, w := range workouts {
				fmt.Printf("%-4d %-15s %-8d %-6d %-6.1f %-12s %s\n", w.ID, w.Exercise, w.Sets, w.Reps, w.Weight, w.Date, w.Notes)
			}
		}
	case "stats":
		total, volume, best, progress := getStatistics(tracker, exercise)
		fmt.Printf("📊 Статистика%s\n", map[bool]string{true: " по упражнению " + exercise, false: ""}[exercise != ""])
		fmt.Printf("Всего тренировок: %d\n", total)
		fmt.Printf("Общий объём: %.1f\n", volume)
		if len(best) > 0 {
			fmt.Println("Лучшие подходы по объёму:")
			for ex, data := range best {
				w := data.Workout
				fmt.Printf("  %s: %d×%d×%.1fкг = %.1f (id %d)\n", ex, w.Sets, w.Reps, w.Weight, data.Volume, w.ID)
			}
		}
		if len(progress) > 0 {
			fmt.Println("Прогресс среднего веса по месяцам:")
			for month, avg := range progress {
				fmt.Printf("  %s: %.1f кг\n", month, avg)
			}
		}
	case "pr":
		records := getPersonalRecords(tracker)
		if len(records) == 0 {
			fmt.Println("Нет рекордов.")
		} else {
			fmt.Println("🏆 Личные рекорды (максимальный объём):")
			for ex, w := range records {
				fmt.Printf("%s: %d×%d×%.1fкг = %.1f (id %d)\n", ex, w.Sets, w.Reps, w.Weight, float64(w.Sets*w.Reps)*w.Weight, w.ID)
			}
		}
	case "delete":
		if id == 0 {
			fmt.Println("Укажите --id")
			return
		}
		if deleteWorkout(tracker, id) {
			fmt.Printf("✅ Тренировка #%d удалена\n", id)
		} else {
			fmt.Printf("❌ Тренировка #%d не найдена\n", id)
		}
	case "export":
		if output == "" {
			fmt.Println("Укажите --output")
			return
		}
		exportCSV(tracker, output)
		fmt.Printf("Экспортировано в %s\n", output)
	default:
		interactiveMode(tracker)
	}
}

func interactiveMode(t *Tracker) {
	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Println("\n🏋️ Дневник тренировок (интерактивный)")
		fmt.Println("1. Добавить тренировку")
		fmt.Println("2. Список тренировок")
		fmt.Println("3. Статистика")
		fmt.Println("4. Личные рекорды")
		fmt.Println("5. Удалить")
		fmt.Println("6. Экспорт CSV")
		fmt.Println("0. Выход")
		fmt.Print("Выберите действие: ")
		scanner.Scan()
		choice := scanner.Text()
		switch choice {
		case "0":
			return
		case "1":
			fmt.Print("Упражнение: ")
			scanner.Scan()
			ex := scanner.Text()
			if ex == "" {
				fmt.Println("Упражнение обязательно")
				continue
			}
			fmt.Print("Подходы: ")
			scanner.Scan()
			sets, _ := strconv.Atoi(scanner.Text())
			fmt.Print("Повторений: ")
			scanner.Scan()
			reps, _ := strconv.Atoi(scanner.Text())
			fmt.Print("Вес (кг): ")
			scanner.Scan()
			weight, _ := strconv.ParseFloat(scanner.Text(), 64)
			fmt.Print("Дата (ГГГГ-ММ-ДД, Enter сегодня): ")
			scanner.Scan()
			date := scanner.Text()
			if date == "" {
				date = time.Now().Format("2006-01-02")
			}
			fmt.Print("Заметки: ")
			scanner.Scan()
			notes := scanner.Text()
			w := addWorkout(t, ex, sets, reps, weight, date, notes)
			fmt.Printf("✅ Добавлена тренировка #%d\n", w.ID)
		case "2":
			fmt.Print("Упражнение (Enter пропустить): ")
			scanner.Scan()
			ex := scanner.Text()
			fmt.Print("Дата от (Enter пропустить): ")
			scanner.Scan()
			from := scanner.Text()
			fmt.Print("Дата до (Enter пропустить): ")
			scanner.Scan()
			to := scanner.Text()
			workouts := getWorkouts(t, ex, from, to, 0, 0)
			if len(workouts) == 0 {
				fmt.Println("Нет записей.")
			} else {
				fmt.Printf("%-4s %-15s %-8s %-6s %-6s %-12s %s\n", "ID", "Упражнение", "Подходы", "Повт.", "Вес", "Дата", "Заметки")
				for _, w := range workouts {
					fmt.Printf("%-4d %-15s %-8d %-6d %-6.1f %-12s %s\n", w.ID, w.Exercise, w.Sets, w.Reps, w.Weight, w.Date, w.Notes)
				}
			}
		case "3":
			fmt.Print("Упражнение (Enter все): ")
			scanner.Scan()
			ex := scanner.Text()
			total, volume, best, progress := getStatistics(t, ex)
			fmt.Printf("📊 Статистика%s\n", map[bool]string{true: " по упражнению " + ex, false: ""}[ex != ""])
			fmt.Printf("Всего тренировок: %d\n", total)
			fmt.Printf("Общий объём: %.1f\n", volume)
			if len(best) > 0 {
				fmt.Println("Лучшие подходы по объёму:")
				for ex, data := range best {
					w := data.Workout
					fmt.Printf("  %s: %d×%d×%.1fкг = %.1f (id %d)\n", ex, w.Sets, w.Reps, w.Weight, data.Volume, w.ID)
				}
			}
			if len(progress) > 0 {
				fmt.Println("Прогресс среднего веса по месяцам:")
				for month, avg := range progress {
					fmt.Printf("  %s: %.1f кг\n", month, avg)
				}
			}
		case "4":
			records := getPersonalRecords(t)
			if len(records) == 0 {
				fmt.Println("Нет рекордов.")
			} else {
				fmt.Println("🏆 Личные рекорды (максимальный объём):")
				for ex, w := range records {
					fmt.Printf("%s: %d×%d×%.1fкг = %.1f (id %d)\n", ex, w.Sets, w.Reps, w.Weight, float64(w.Sets*w.Reps)*w.Weight, w.ID)
				}
			}
		case "5":
			fmt.Print("ID для удаления: ")
			scanner.Scan()
			idStr := scanner.Text()
			id, _ := strconv.Atoi(idStr)
			if deleteWorkout(t, id) {
				fmt.Println("✅ Удалено")
			} else {
				fmt.Println("❌ Не найдено")
			}
		case "6":
			fmt.Print("Имя файла (CSV): ")
			scanner.Scan()
			fname := scanner.Text()
			if fname == "" {
				fname = "workouts.csv"
			}
			exportCSV(t, fname)
			fmt.Printf("Экспортировано в %s\n", fname)
		default:
			fmt.Println("Неверный выбор")
		}
	}
}

File: workout.rs
// workout.rs - Дневник тренировок на Rust (CLI)
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::fs;
use std::io::{self, Write, BufRead};
use std::path::Path;
use std::str::FromStr;
use chrono::Local;

#[derive(Serialize, Deserialize, Clone)]
struct Workout {
    id: u32,
    exercise: String,
    sets: u32,
    reps: u32,
    weight: f64,
    date: String,
    notes: String,
}

#[derive(Serialize, Deserialize)]
struct Tracker {
    workouts: Vec<Workout>,
    next_id: u32,
}

impl Tracker {
    fn load() -> Self {
        let path = "workouts.json";
        if Path::new(path).exists() {
            if let Ok(data) = fs::read_to_string(path) {
                if let Ok(t) = serde_json::from_str(&data) {
                    return t;
                }
            }
        }
        Tracker { workouts: vec![], next_id: 1 }
    }

    fn save(&self) {
        let data = serde_json::to_string_pretty(self).unwrap();
        fs::write("workouts.json", data).unwrap();
    }
}

fn add_workout(tracker: &mut Tracker, exercise: &str, sets: u32, reps: u32, weight: f64,
               date: &str, notes: &str) -> Workout {
    let date = if date.is_empty() { Local::now().format("%Y-%m-%d").to_string() } else { date.to_string() };
    let w = Workout {
        id: tracker.next_id,
        exercise: exercise.to_string(),
        sets,
        reps,
        weight,
        date,
        notes: notes.to_string(),
    };
    tracker.workouts.push(w.clone());
    tracker.next_id += 1;
    tracker.save();
    w
}

fn delete_workout(tracker: &mut Tracker, id: u32) -> bool {
    let len = tracker.workouts.len();
    tracker.workouts.retain(|w| w.id != id);
    if tracker.workouts.len() < len {
        tracker.save();
        return true;
    }
    false
}

fn get_workouts(tracker: &Tracker, exercise: Option<&str>, date_from: Option<&str>, date_to: Option<&str>,
                min_weight: Option<f64>, max_weight: Option<f64>) -> Vec<Workout> {
    let mut res = tracker.workouts.clone();
    if let Some(ex) = exercise {
        res.retain(|w| w.exercise.to_lowercase() == ex.to_lowercase());
    }
    if let Some(d) = date_from {
        res.retain(|w| w.date >= d);
    }
    if let Some(d) = date_to {
        res.retain(|w| w.date <= d);
    }
    if let Some(m) = min_weight {
        res.retain(|w| w.weight >= m);
    }
    if let Some(m) = max_weight {
        res.retain(|w| w.weight <= m);
    }
    res.sort_by(|a, b| a.date.cmp(&b.date));
    res
}

fn get_statistics(tracker: &Tracker, exercise: Option<&str>) -> (u32, f64, HashMap<String, (Workout, f64)>, HashMap<String, f64>) {
    let workouts = get_workouts(tracker, exercise, None, None, None, None);
    let total_workouts = workouts.len() as u32;
    let total_volume = workouts.iter().map(|w| w.sets as f64 * w.reps as f64 * w.weight).sum();
    let mut best_by_exercise = HashMap::new();
    for w in &workouts {
        let vol = w.sets as f64 * w.reps as f64 * w.weight;
        let entry = best_by_exercise.entry(w.exercise.clone()).or_insert((w.clone(), 0.0));
        if vol > entry.1 {
            *entry = (w.clone(), vol);
        }
    }
    let mut progress = HashMap::new();
    if exercise.is_some() {
        let mut month_map = HashMap::new();
        for w in &workouts {
            let month = w.date[..7].to_string();
            let entry = month_map.entry(month).or_insert((0.0, 0));
            entry.0 += w.weight;
            entry.1 += 1;
        }
        let mut months: Vec<_> = month_map.keys().collect();
        months.sort();
        for m in months {
            let (total, count) = month_map[m];
            progress.insert(m.clone(), total / count as f64);
        }
    }
    (total_workouts, total_volume, best_by_exercise, progress)
}

fn get_personal_records(tracker: &Tracker) -> HashMap<String, Workout> {
    let mut best = HashMap::new();
    for w in &tracker.workouts {
        let vol = w.sets as f64 * w.reps as f64 * w.weight;
        let entry = best.entry(w.exercise.clone()).or_insert(w.clone());
        if vol > (entry.sets as f64 * entry.reps as f64 * entry.weight) {
            *entry = w.clone();
        }
    }
    best
}

fn export_csv(tracker: &Tracker, filepath: &str) -> Result<(), Box<dyn std::error::Error>> {
    let mut writer = csv::Writer::from_path(filepath)?;
    writer.write_record(&["ID", "Exercise", "Sets", "Reps", "Weight", "Date", "Notes"])?;
    for w in &tracker.workouts {
        writer.serialize((w.id, &w.exercise, w.sets, w.reps, w.weight, &w.date, &w.notes))?;
    }
    writer.flush()?;
    Ok(())
}

fn read_line(prompt: &str) -> String {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() < 2 {
        interactive_mode();
        return;
    }
    let mut tracker = Tracker::load();
    match args[1].as_str() {
        "add" => {
            let mut exercise = String::new();
            let mut sets = 0;
            let mut reps = 0;
            let mut weight = 0.0;
            let mut date = String::new();
            let mut notes = String::new();
            let mut i = 2;
            while i < args.len() {
                match args[i].as_str() {
                    "--exercise" => { exercise = args[i+1].clone(); i += 2; }
                    "--sets" => { sets = args[i+1].parse().unwrap_or(0); i += 2; }
                    "--reps" => { reps = args[i+1].parse().unwrap_or(0); i += 2; }
                    "--weight" => { weight = args[i+1].parse().unwrap_or(0.0); i += 2; }
                    "--date" => { date = args[i+1].clone(); i += 2; }
                    "--notes" => { notes = args[i+1].clone(); i += 2; }
                    _ => { i += 1; }
                }
            }
            if exercise.is_empty() || sets == 0 || reps == 0 || weight == 0.0 {
                println!("Укажите --exercise, --sets, --reps, --weight");
                return;
            }
            let w = add_workout(&mut tracker, &exercise, sets, reps, weight, &date, &notes);
            println!("✅ Тренировка #{} добавлена: {} {}×{}×{:.1}кг", w.id, w.exercise, w.sets, w.reps, w.weight);
        }
        "list" => {
            let mut exercise = None;
            let mut date_from = None;
            let mut date_to = None;
            let mut i = 2;
            while i < args.len() {
                match args[i].as_str() {
                    "--exercise" => { exercise = Some(args[i+1].clone()); i += 2; }
                    "--from" => { date_from = Some(args[i+1].clone()); i += 2; }
                    "--to" => { date_to = Some(args[i+1].clone()); i += 2; }
                    _ => { i += 1; }
                }
            }
            let workouts = get_workouts(&tracker, exercise.as_deref(), date_from.as_deref(), date_to.as_deref(), None, None);
            if workouts.is_empty() {
                println!("Нет записей.");
            } else {
                println!("{:<4} {:<15} {:<8} {:<6} {:<6} {:<12} {}", "ID", "Упражнение", "Подходы", "Повт.", "Вес", "Дата", "Заметки");
                for w in workouts {
                    println!("{:<4} {:<15} {:<8} {:<6} {:<6.1} {:<12} {}", w.id, w.exercise, w.sets, w.reps, w.weight, w.date, w.notes);
                }
            }
        }
        "stats" => {
            let mut exercise = None;
            let mut i = 2;
            while i < args.len() {
                if args[i] == "--exercise" {
                    exercise = Some(args[i+1].clone());
                    break;
                }
                i += 1;
            }
            let (total, volume, best, progress) = get_statistics(&tracker, exercise.as_deref());
            println!("📊 Статистика{}", if exercise.is_some() { format!(" по упражнению {}", exercise.unwrap()) } else { String::new() });
            println!("Всего тренировок: {}", total);
            println!("Общий объём: {:.1}", volume);
            if !best.is_empty() {
                println!("Лучшие подходы по объёму:");
                for (ex, (w, vol)) in best {
                    println!("  {}: {}×{}×{:.1}кг = {:.1} (id {})", ex, w.sets, w.reps, w.weight, vol, w.id);
                }
            }
            if !progress.is_empty() {
                println!("Прогресс среднего веса по месяцам:");
                let mut months: Vec<_> = progress.keys().collect();
                months.sort();
                for m in months {
                    println!("  {}: {:.1} кг", m, progress[m]);
                }
            }
        }
        "pr" => {
            let records = get_personal_records(&tracker);
            if records.is_empty() {
                println!("Нет рекордов.");
            } else {
                println!("🏆 Личные рекорды (максимальный объём):");
                for (ex, w) in records {
                    let vol = w.sets as f64 * w.reps as f64 * w.weight;
                    println!("{}: {}×{}×{:.1}кг = {:.1} (id {})", ex, w.sets, w.reps, w.weight, vol, w.id);
                }
            }
        }
        "delete" => {
            let mut id = 0;
            let mut i = 2;
            while i < args.len() {
                if args[i] == "--id" {
                    id = args[i+1].parse().unwrap_or(0);
                    break;
                }
                i += 1;
            }
            if id == 0 {
                println!("Укажите --id");
                return;
            }
            if delete_workout(&mut tracker, id) {
                println!("✅ Тренировка #{} удалена", id);
            } else {
                println!("❌ Тренировка #{} не найдена", id);
            }
        }
        "export" => {
            let mut output = String::new();
            let mut i = 2;
            while i < args.len() {
                if args[i] == "--output" {
                    output = args[i+1].clone();
                    break;
                }
                i += 1;
            }
            if output.is_empty() {
                println!("Укажите --output");
                return;
            }
            if let Err(e) = export_csv(&tracker, &output) {
                println!("Ошибка экспорта: {}", e);
            } else {
                println!("Экспортировано в {}", output);
            }
        }
        _ => interactive_mode(),
    }
}

fn interactive_mode() {
    let mut tracker = Tracker::load();
    let stdin = io::stdin();
    let mut stdout = io::stdout();
    loop {
        println!("\n🏋️ Дневник тренировок (интерактивный)");
        println!("1. Добавить тренировку");
        println!("2. Список тренировок");
        println!("3. Статистика");
        println!("4. Личные рекорды");
        println!("5. Удалить");
        println!("6. Экспорт CSV");
        println!("0. Выход");
        print!("Выберите действие: ");
        stdout.flush().unwrap();
        let mut choice = String::new();
        stdin.read_line(&mut choice).unwrap();
        match choice.trim() {
            "0" => break,
            "1" => {
                print!("Упражнение: ");
                stdout.flush().unwrap();
                let mut ex = String::new();
                stdin.read_line(&mut ex).unwrap();
                let ex = ex.trim();
                if ex.is_empty() {
                    println!("Упражнение обязательно");
                    continue;
                }
                print!("Подходы: ");
                stdout.flush().unwrap();
                let mut sets_str = String::new();
                stdin.read_line(&mut sets_str).unwrap();
                let sets = sets_str.trim().parse::<u32>().unwrap_or(0);
                print!("Повторений: ");
                stdout.flush().unwrap();
                let mut reps_str = String::new();
                stdin.read_line(&mut reps_str).unwrap();
                let reps = reps_str.trim().parse::<u32>().unwrap_or(0);
                print!("Вес (кг): ");
                stdout.flush().unwrap();
                let mut weight_str = String::new();
                stdin.read_line(&mut weight_str).unwrap();
                let weight = weight_str.trim().parse::<f64>().unwrap_or(0.0);
                print!("Дата (ГГГГ-ММ-ДД, Enter сегодня): ");
                stdout.flush().unwrap();
                let mut date = String::new();
                stdin.read_line(&mut date).unwrap();
                let date = date.trim();
                print!("Заметки: ");
                stdout.flush().unwrap();
                let mut notes = String::new();
                stdin.read_line(&mut notes).unwrap();
                let w = add_workout(&mut tracker, ex, sets, reps, weight, date, notes.trim());
                println!("✅ Добавлена тренировка #{}", w.id);
            }
            "2" => {
                print!("Упражнение (Enter пропустить): ");
                stdout.flush().unwrap();
                let mut ex = String::new();
                stdin.read_line(&mut ex).unwrap();
                let ex = if ex.trim().is_empty() { None } else { Some(ex.trim()) };
                print!("Дата от (Enter пропустить): ");
                stdout.flush().unwrap();
                let mut from = String::new();
                stdin.read_line(&mut from).unwrap();
                let from = if from.trim().is_empty() { None } else { Some(from.trim()) };
                print!("Дата до (Enter пропустить): ");
                stdout.flush().unwrap();
                let mut to = String::new();
                stdin.read_line(&mut to).unwrap();
                let to = if to.trim().is_empty() { None } else { Some(to.trim()) };
                let workouts = get_workouts(&tracker, ex, from, to, None, None);
                if workouts.is_empty() {
                    println!("Нет записей.");
                } else {
                    println!("{:<4} {:<15} {:<8} {:<6} {:<6} {:<12} {}", "ID", "Упражнение", "Подходы", "Повт.", "Вес", "Дата", "Заметки");
                    for w in workouts {
                        println!("{:<4} {:<15} {:<8} {:<6} {:<6.1} {:<12} {}", w.id, w.exercise, w.sets, w.reps, w.weight, w.date, w.notes);
                    }
                }
            }
            "3" => {
                print!("Упражнение (Enter все): ");
                stdout.flush().unwrap();
                let mut ex = String::new();
                stdin.read_line(&mut ex).unwrap();
                let ex = if ex.trim().is_empty() { None } else { Some(ex.trim()) };
                let (total, volume, best, progress) = get_statistics(&tracker, ex);
                println!("📊 Статистика{}", if ex.is_some() { format!(" по упражнению {}", ex.unwrap()) } else { String::new() });
                println!("Всего тренировок: {}", total);
                println!("Общий объём: {:.1}", volume);
                if !best.is_empty() {
                    println!("Лучшие подходы по объёму:");
                    for (ex, (w, vol)) in best {
                        println!("  {}: {}×{}×{:.1}кг = {:.1} (id {})", ex, w.sets, w.reps, w.weight, vol, w.id);
                    }
                }
                if !progress.is_empty() {
                    println!("Прогресс среднего веса по месяцам:");
                    let mut months: Vec<_> = progress.keys().collect();
                    months.sort();
                    for m in months {
                        println!("  {}: {:.1} кг", m, progress[m]);
                    }
                }
            }
            "4" => {
                let records = get_personal_records(&tracker);
                if records.is_empty() {
                    println!("Нет рекордов.");
                } else {
                    println!("🏆 Личные рекорды (максимальный объём):");
                    for (ex, w) in records {
                        let vol = w.sets as f64 * w.reps as f64 * w.weight;
                        println!("{}: {}×{}×{:.1}кг = {:.1} (id {})", ex, w.sets, w.reps, w.weight, vol, w.id);
                    }
                }
            }
            "5" => {
                print!("ID для удаления: ");
                stdout.flush().unwrap();
                let mut id_str = String::new();
                stdin.read_line(&mut id_str).unwrap();
                let id = id_str.trim().parse::<u32>().unwrap_or(0);
                if id == 0 {
                    println!("Неверный ID");
                    continue;
                }
                if delete_workout(&mut tracker, id) {
                    println!("✅ Удалено");
                } else {
                    println!("❌ Не найдено");
                }
            }
            "6" => {
                print!("Имя файла (CSV): ");
                stdout.flush().unwrap();
                let mut fname = String::new();
                stdin.read_line(&mut fname).unwrap();
                let fname = if fname.trim().is_empty() { "workouts.csv".to_string() } else { fname.trim().to_string() };
                if let Err(e) = export_csv(&tracker, &fname) {
                    println!("Ошибка экспорта: {}", e);
                } else {
                    println!("Экспортировано в {}", fname);
                }
            }
            _ => println!("Неверный выбор"),
        }
    }
}
